Wikipedia/Networking/Fetchers/SavedArticlesFetcher.m

```diff
@@ -27,59 +24,188 @@ @implementation SavedArticlesFetcher
 
 #pragma mark - Shared Access
 
-static SavedArticlesFetcher* _fetcher = nil;
+static SavedArticlesFetcher* _articleFetcher = nil;
 
 + (SavedArticlesFetcher*)sharedInstance {
-    return _fetcher;
+    NSParameterAssert([NSThread isMainThread]);
+    return _articleFetcher;
 }
 
-+ (void)setSharedInstance:(SavedArticlesFetcher*)fetcher {
-    _fetcher = fetcher;
++ (void)setSharedInstance:(SavedArticlesFetcher*)articleFetcher {
+    NSParameterAssert([NSThread isMainThread]);
+    _articleFetcher = articleFetcher;
 }
 
-- (instancetype)initWithDataStore:(MWKDataStore*)dataStore {
-    NSParameterAssert(dataStore != nil);
-
+- (instancetype)initWithArticleFetcher:(WMFArticleFetcher*)articleFetcher
+                       imageController:(WMFImageController*)imageController {
+    NSParameterAssert(articleFetcher);
     self = [super init];
     if (self) {
-        self.dataStore   = dataStore;
-        self.accessQueue = dispatch_queue_create("org.wikipedia.savedarticlesfetcher.accessQueue", DISPATCH_QUEUE_SERIAL);
-        self.fetcher     = [[WMFArticleFetcher alloc] initWithDataStore:self.dataStore];
+        _accessQueue                       = dispatch_queue_create("org.wikipedia.savedarticlesarticleFetcher.accessQueue", DISPATCH_QUEUE_SERIAL);
+        self.fetchOperationsByArticleTitle = [NSMutableDictionary new];
+        self.errorsByArticleTitle          = [NSMutableDictionary new];
+        self.articleFetcher                = articleFetcher;
+        self.imageController               = imageController;
     }
     return self;
 }
 
-- (void)fetchSavedPageList:(MWKSavedPageList*)savedPageList {
-    self.savedPageList = savedPageList;
+- (instancetype)initWithDataStore:(MWKDataStore*)dataStore {
+    return [self initWithArticleFetcher:[[WMFArticleFetcher alloc] initWithDataStore:dataStore]
+                        imageController:[WMFImageController sharedInstance]];
+}
+
+- (void)setSavedPageList:(MWKSavedPageList*)savedPageList {
```

Comment:
This should also get removed since we don't need a setter to do all this book keeping since the saved page list should never change. 

This back to Immutability - why would we make this mutable and add the additional book keeping when we know in practice we will never change this property.